finagle/buoyant/src/main/scala/com/twitter/finagle/buoyant/ExistentialStability.scala

```diff
@@ -56,44 +69,54 @@ object ExistentialStability {
     extends AnyVal {
     def stabilizeExistence: Activity[Option[Var[T]]] = {
-      val inner = Var.async[Activity.State[Option[VarUp[T]]]](Activity.Pending) { update =>
-        // the current inner Var, null if the outer Var is None
-        @volatile var current: VarUp[T] = null
-        @volatile var exists = false
-        val mu = new {}
-        unstable.states.respond {
-          case Activity.Ok(Some(t)) if current == null => mu.synchronized {
-            // T created
-            current = Var(t)
-            exists = true
-            update() = Activity.Ok(Some(current))
-          }
-          case Activity.Ok(Some(t)) if !exists => mu.synchronized {
-            // T recreated
-            exists = true
-            current() = t
-            update() = Activity.Ok(Some(current))
-          }
-          case Activity.Ok(Some(t)) => mu.synchronized {
-            // T modified
-            current() = t
-          }
-          case Activity.Ok(None) => mu.synchronized {
-            // T deleted
-            exists = false
-            update() = Activity.Ok(None)
-          }
-          case Activity.Pending => mu.synchronized {
-            update() = Activity.Pending
-            exists = false
-          }
-          case Activity.Failed(e) => mu.synchronized {
-            update() = Activity.Failed(e)
-            exists = false
+
+      // The inner Var is null until the first time unstable becomes a Some.
+      @volatile var inner: Var[T] = null
+
+      // This keeps track of the state of the outer Var:
+      // None means the state was not Activity.Ok
+      // Some(false) means the state was Activity.Ok(Some)
+      // Some(true) means the state was Activity.Ok(None)
+      //
+      // We need to track these three states because we only want to update the outer Var when we
+      // transition between states.
+      @volatile var exists: Option[Boolean] = None
+      val mu = new {}
+
+      val outer = Var.async[Activity.State[Option[Var[T]]]](Activity.Pending) { update =>
+        unstable.states.respond { state =>
+          mu.synchronized {
+            state match {
+              case Activity.Ok(Some(_)) if exists == Some(true) =>
+              // Do nothing.
+              case Activity.Ok(Some(t)) if exists != Some(true) =>
+                exists = Some(true)
+                if (inner == null) {
+                  inner = Var.async(t) { update =>
+                    unstable.states.respond {
+                      case Activity.Ok(Some(tt)) => update() = tt
+                      case _ =>
+                    }
+                  }
+                }
+                update() = Activity.Ok(Some(inner))
+              case Activity.Ok(None) if exists == Some(false) =>
+              // Do nothing.
+              case Activity.Ok(None) if exists != Some(false) =>
+                exists = Some(false)
+                update() = Activity.Ok(None)
+              case Activity.Pending =>
```

Comment:
Nit/TIOLI: maybe we should match `Pending` and `Failed` in the same match arm, like this:
```scala
case state =>
  exists = None
  update() = state
```
That makes the code a little less repetitive.

On the other hand, this does rely on the rest of the match exhaustively matching over all the possible `Activity.Ok(...)` patterns...which it _does_, but it might be possible to miss one in a future change... 🤷‍♀️

Reply:
This actually doesn't work because state is of type `Activity.State[Option[T]]` and update expects type `Activity.State[Option[Var[T]]`.

I bet you might be able to pull this off with Scala3's union types because `Activity.Pending | Activity.Failed` is a subtype of `Activity.State[Nothing]` but I don't think this can be done in scala2.

Reply:
Turns out I totally forgot how pattern matching works in Scala. Man, languages with subtyping are _wild_...